xarray/core/dataset.py

```diff
@@ -2416,6 +2416,60 @@ def from_dataframe(cls, dataframe):
             obj[name] = (dims, data)
         return obj
 
+    def to_dask_dataframe(self, set_index=False):
+        """
+        Convert this dataset into a dask.dataframe.DataFrame.
+
+        Both the data and coordinate variables in this dataset form the columns of the
+        DataFrame.
+
+        If set_index=True, the dask DataFrame is indexed by
+        this dataset's coordinate. Since dask DataFrames to not support
+        multi-indexes, set_index only works if there is one coordinate dimension.
+        """
+
+        import dask.dataframe as dd
+
+        ordered_dims = self.dims
+        coord_columns = [c for c in self if c in ordered_dims]
+
+        # convert all coordinates into data variables
+        ds = self.reset_index(coord_columns).reset_coords()
+
+        # ensure we are dealing with dask arrays with consistent chunks
+        chunked = ds.chunk(chunks=self.chunks)
+
+        # order columns so that coordinates appear before data
+        columns = list(chunked)
+        num_coords = len(coord_columns)
+        columns = columns[-num_coords:] + columns[:-num_coords]
+
+        data = [chunked._variables[k].set_dims(ordered_dims).data.reshape(-1)
+                for k in columns]
+
+        # restore names of columns that were also index names (e.g. x_ -> x)
+        columns = [c[:-1] if c[:-1] in ds.dims else c for c in columns]
```

Comment:
This feels a little fragile -- it would be nice to avoid if possible. (See my note above on avoiding `reset_index()`/`reset_coords()`)

Reply:
I've restructured the code so that this is no longer needed.